File: csharp/src/EcfClient/Comprobantes/Validation.cs
using System.Text.RegularExpressions;
using EcfClient.Comprobantes.Datos;
using EcfClient.Exceptions;

namespace EcfClient.Comprobantes;

internal static class Validation
{
    private static readonly Regex FechaDdMmYyyy = new(@"^\d{2}-\d{2}-\d{4}$", RegexOptions.Compiled);

    private static bool IsFechaDdMmYyyy(string? s)
    {
        if (string.IsNullOrWhiteSpace(s) || !FechaDdMmYyyy.Match(s.Trim()).Success) return false;
        var parts = s.Trim().Split('-');
        if (parts.Length != 3) return false;
        if (!int.TryParse(parts[0], out var d) || !int.TryParse(parts[1], out var m) || !int.TryParse(parts[2], out var y))
            return false;
        return m >= 1 && m <= 12 && d >= 1 && d <= 31 && y >= 1900 && y <= 2100;
    }

    private static bool IsRncCedulaValid(string? val)
    {
        if (string.IsNullOrWhiteSpace(val)) return false;
        var digits = Regex.Replace(val, @"\D", "");
        return digits.Length is 9 or 11;
    }

    internal static void ValidarFecha(string path, string? valor, bool obligatorio = true)
    {
        if (string.IsNullOrWhiteSpace(valor))
        {
            if (obligatorio)
                throw new EcfValidationException($"Campo requerido: {path}", new[] { new ValidationError(path, "ecf.requerido") });
            return;
        }
        if (!IsFechaDdMmYyyy(valor))
            throw new EcfValidationException($"Formato fecha inválido (dd-MM-yyyy): {path}",
                new[] { new ValidationError(path, "ecf.formato_fecha", new Dictionary<string, object?> { ["valor"] = valor }) });
    }

    internal static void ValidarRnc(string path, string? valor, bool obligatorio = true)
    {
        if (string.IsNullOrWhiteSpace(valor))
        {
            if (obligatorio)
                throw new EcfValidationException($"Campo requerido: {path}", new[] { new ValidationError(path, "ecf.requerido") });
            return;
        }
        if (!IsRncCedulaValid(valor))
            throw new EcfValidationException($"RNC o cédula con formato inválido: {path}", new[] { new ValidationError(path, "ecf.formato_rnc") });
    }

    internal static void ValidarEncf(string? encf)
    {
        if (string.IsNullOrWhiteSpace(encf))
            throw new EcfValidationException("eNCF requerido", new[] { new ValidationError("Encabezado.IdDoc.eNCF", "ecf.requerido") });
        if (encf.Trim().Length > 13)
            throw new EcfValidationException("eNCF longitud máxima 13", new[] { new ValidationError("Encabezado.IdDoc.eNCF", "ecf.largo_maximo", new Dictionary<string, object?> { ["max"] = 13 }) });
    }

    internal static void ValidarEmisor(EmisorData emisor, bool exigirDireccion = true)
    {
        ValidarRnc("Encabezado.Emisor.RNCEmisor", emisor.Rnc);
        if (string.IsNullOrWhiteSpace(emisor.RazonSocial))
            throw new EcfValidationException("RazonSocialEmisor requerido", new[] { new ValidationError("Encabezado.Emisor.RazonSocialEmisor", "ecf.requerido") });
        if (emisor.RazonSocial.Trim().Length > 150)
            throw new EcfValidationException("RazonSocialEmisor máximo 150 caracteres", new[] { new ValidationError("Encabezado.Emisor.RazonSocialEmisor", "ecf.largo_maximo", new Dictionary<string, object?> { ["max"] = 150 }) });
        if (exigirDireccion && string.IsNullOrWhiteSpace(emisor.Direccion))
            throw new EcfValidationException("DireccionEmisor requerido", new[] { new ValidationError("Encabezado.Emisor.DireccionEmisor", "ecf.requerido") });
        if (!string.IsNullOrWhiteSpace(emisor.Direccion) && emisor.Direccion.Trim().Length > 100)
            throw new EcfValidationException("DireccionEmisor máximo 100 caracteres", new[] { new ValidationError("Encabezado.Emisor.DireccionEmisor", "ecf.largo_maximo", new Dictionary<string, object?> { ["max"] = 100 }) });
        ValidarFecha("Encabezado.Emisor.FechaEmision", emisor.FechaEmision);
    }

    internal static void ValidarComprador(CompradorData comprador, bool obligatorioComprador = true)
    {
        if (string.IsNullOrWhiteSpace(comprador.RazonSocial))
        {
            if (obligatorioComprador)
                throw new EcfValidationException("RazonSocialComprador requerido", new[] { new ValidationError("Encabezado.Comprador.RazonSocialComprador", "ecf.requerido") });
            return;
        }
        if (!string.IsNullOrWhiteSpace(comprador.Rnc) && !IsRncCedulaValid(comprador.Rnc))
            throw new EcfValidationException("RNCComprador formato inválido", new[] { new ValidationError("Encabezado.Comprador.RNCComprador", "ecf.formato_rnc") });
        var rncOk = (comprador.Rnc ?? "").Trim();
        var idExt = (comprador.IdentificadorExtranjero ?? "").Trim();
        if (!string.IsNullOrWhiteSpace(comprador.RazonSocial) && string.IsNullOrEmpty(rncOk) && string.IsNullOrEmpty(idExt))
            throw new EcfValidationException("RNC o IdentificadorExtranjero requerido cuando hay Comprador", new[] { new ValidationError("Encabezado.Comprador.RNCComprador", "ecf.requerido") });
    }

    internal static void ValidarItems(IList<DetalleItem> items, int tipoEcf)
    {
        if (items == null || items.Count == 0)
            throw new EcfValidationException("Al menos un ítem en DetallesItems", new[] { new ValidationError("DetallesItems", "ecf.detalle_al_menos_uno") });
        for (var i = 0; i < items.Count; i++)
        {
            var it = items[i];
            var prefix = $"DetallesItems[{i}]";
            if (it.NumeroLinea < 1 || it.NumeroLinea > 10000)
                throw new EcfValidationException($"NumeroLinea 1-10000: {prefix}", new[] { new ValidationError($"{prefix}.NumeroLinea", "ecf.rango_numero_linea") });
            if (string.IsNullOrWhiteSpace(it.NombreItem))
                throw new EcfValidationException($"NombreItem requerido: {prefix}", new[] { new ValidationError($"{prefix}.NombreItem", "ecf.requerido") });
            if (it.NombreItem.Trim().Length > 80)
                throw new EcfValidationException($"NombreItem máximo 80: {prefix}", new[] { new ValidationError($"{prefix}.NombreItem", "ecf.largo_maximo") });
            if (it.IndicadorFacturacion < 0 || it.IndicadorFacturacion > 4)
                throw new EcfValidationException($"IndicadorFacturacion 0-4: {prefix}", new[] { new ValidationError($"{prefix}.IndicadorFacturacion", "ecf.valor_no_permitido") });
            if (tipoEcf is 43 or 44 or 47 && it.IndicadorFacturacion != 4)
                throw new EcfValidationException($"Tipo {tipoEcf} requiere IndicadorFacturacion 4: {prefix}", new[] { new ValidationError($"{prefix}.IndicadorFacturacion", "ecf.indicador_facturacion_tipo_ecf") });
            if (tipoEcf == 46 && it.IndicadorFacturacion != 3)
                throw new EcfValidationException("Tipo 46 requiere IndicadorFacturacion 3: " + prefix, new[] { new ValidationError($"{prefix}.IndicadorFacturacion", "ecf.indicador_facturacion_tipo_ecf") });
            if (it.IndicadorBienServicio is not 1 and not 2)
                throw new EcfValidationException($"IndicadorBienoServicio 1 o 2: {prefix}", new[] { new ValidationError($"{prefix}.IndicadorBienoServicio", "ecf.valor_no_permitido") });
            if (tipoEcf == 47 && it.IndicadorBienServicio != 2)
                throw new EcfValidationException("Tipo 47 requiere IndicadorBienoServicio 2: " + prefix, new[] { new ValidationError($"{prefix}.IndicadorBienoServicio", "ecf.indicador_bien_servicio_tipo_47") });
            if (it.Cantidad <= 0)
                throw new EcfValidationException($"CantidadItem > 0: {prefix}", new[] { new ValidationError($"{prefix}.CantidadItem", "ecf.mayor_cero") });
            if (it.PrecioUnitario < 0 || it.MontoItem < 0)
                throw new EcfValidationException($"PrecioUnitario/MontoItem no negativos: {prefix}", new[] { new ValidationError($"{prefix}.MontoItem", "ecf.valor_no_negativo") });
        }
    }

    internal static void ValidarInformacionReferencia(InformacionReferencia info)
    {
        if (string.IsNullOrWhiteSpace(info.NcfModificado))
            throw new EcfValidationException("NCFModificado requerido para tipo 33/34", new[] { new ValidationError("InformacionReferencia.NCFModificado", "ecf.requerido") });
        ValidarFecha("InformacionReferencia.FechaNCFModificado", info.FechaNcfModificado);
        var cod = (info.CodigoModificacion ?? "").Trim();
        if (string.IsNullOrEmpty(cod) || (cod != "1" && cod != "2" && cod != "3" && cod != "4"))
            throw new EcfValidationException("CodigoModificacion debe ser 1, 2, 3 o 4", new[] { new ValidationError("InformacionReferencia.CodigoModificacion", "ecf.valor_no_permitido") });
    }
}

File: python-driver/src/chalona_driver/comprobantes/validation.py
"""Validación de comprobantes e-CF según manual DGII. Lanza EcfValidationError."""

import re

from chalona_driver.exceptions import EcfValidationError

_FECHA_DD_MM_YYYY = re.compile(r"^\d{2}-\d{2}-\d{4}$")


def _is_fecha_dd_mm_yyyy(s: str) -> bool:
    if not s or not _FECHA_DD_MM_YYYY.match(s.strip()):
        return False
    parts = s.strip().split("-")
    d, m, y = int(parts[0]), int(parts[1]), int(parts[2])
    if m < 1 or m > 12 or d < 1 or d > 31 or y < 1900 or y > 2100:
        return False
    return True


def _is_rnc_cedula_valid(val: str) -> bool:
    """RNC o cédula RD: solo dígitos, 9 u 11 caracteres."""
    if not val:
        return False
    digits = re.sub(r"\D", "", val)
    return len(digits) in (9, 11)


def validar_fecha(path: str, valor: str, obligatorio: bool = True) -> None:
    if not valor or not valor.strip():
        if obligatorio:
            raise EcfValidationError(f"Campo requerido: {path}", [{"path": path, "code": "ecf.requerido"}])
        return
    if not _is_fecha_dd_mm_yyyy(valor):
        raise EcfValidationError(
            f"Formato fecha inválido (dd-MM-yyyy): {path}",
            [{"path": path, "code": "ecf.formato_fecha", "valor": valor}],
        )


def validar_rnc(path: str, valor: str, obligatorio: bool = True) -> None:
    if not valor or not valor.strip():
        if obligatorio:
            raise EcfValidationError(f"Campo requerido: {path}", [{"path": path, "code": "ecf.requerido"}])
        return
    if not _is_rnc_cedula_valid(valor):
        raise EcfValidationError(
            f"RNC o cédula con formato inválido: {path}",
            [{"path": path, "code": "ecf.formato_rnc"}],
        )


def validar_encf(encf: str) -> None:
    if not encf or not encf.strip():
        raise EcfValidationError(
            "eNCF requerido",
            [{"path": "Encabezado.IdDoc.eNCF", "code": "ecf.requerido"}],
        )
    if len(encf.strip()) > 13:
        raise EcfValidationError(
            "eNCF longitud máxima 13",
            [{"path": "Encabezado.IdDoc.eNCF", "code": "ecf.largo_maximo", "max": 13}],
        )


def validar_emisor(emisor: "EmisorData", exigir_direccion: bool = True) -> None:
    validar_rnc("Encabezado.Emisor.RNCEmisor", emisor.rnc)
    if not emisor.razon_social or not emisor.razon_social.strip():
        raise EcfValidationError(
            "RazonSocialEmisor requerido",
            [{"path": "Encabezado.Emisor.RazonSocialEmisor", "code": "ecf.requerido"}],
        )
    if len(emisor.razon_social.strip()) > 150:
        raise EcfValidationError(
            "RazonSocialEmisor máximo 150 caracteres",
            [{"path": "Encabezado.Emisor.RazonSocialEmisor", "code": "ecf.largo_maximo", "max": 150}],
        )
    if exigir_direccion and (not emisor.direccion or not emisor.direccion.strip()):
        raise EcfValidationError(
            "DireccionEmisor requerido",
            [{"path": "Encabezado.Emisor.DireccionEmisor", "code": "ecf.requerido"}],
        )
    if emisor.direccion and len(emisor.direccion.strip()) > 100:
        raise EcfValidationError(
            "DireccionEmisor máximo 100 caracteres",
            [{"path": "Encabezado.Emisor.DireccionEmisor", "code": "ecf.largo_maximo", "max": 100}],
        )
    validar_fecha("Encabezado.Emisor.FechaEmision", emisor.fecha_emision)


def validar_comprador(comprador: "CompradorData", obligatorio_comprador: bool = True) -> None:
    if not comprador.razon_social or not comprador.razon_social.strip():
        if obligatorio_comprador:
            raise EcfValidationError(
                "RazonSocialComprador requerido",
                [{"path": "Encabezado.Comprador.RazonSocialComprador", "code": "ecf.requerido"}],
            )
        return
    if comprador.rnc and not _is_rnc_cedula_valid(comprador.rnc):
        raise EcfValidationError(
            "RNCComprador formato inválido",
            [{"path": "Encabezado.Comprador.RNCComprador", "code": "ecf.formato_rnc"}],
        )
    rnc_ok = (comprador.rnc or "").strip()
    id_ext = (comprador.identificador_extranjero or "").strip()
    if comprador.razon_social.strip() and not rnc_ok and not id_ext:
        raise EcfValidationError(
            "RNC o IdentificadorExtranjero requerido cuando hay Comprador",
            [{"path": "Encabezado.Comprador.RNCComprador", "code": "ecf.requerido"}],
        )


def validar_items(items: list["DetalleItem"], tipo_ecf: int) -> None:
    from .datos import DetalleItem

    if not items:
        raise EcfValidationError(
            "Al menos un ítem en DetallesItems",
            [{"path": "DetallesItems", "code": "ecf.detalle_al_menos_uno"}],
        )
    indicadores_fact = (0, 1, 2, 3, 4)
    indicadores_bien = (1, 2)
    for i, it in enumerate(items):
        prefix = f"DetallesItems[{i}]"
        if it.numero_linea < 1 or it.numero_linea > 10000:
            raise EcfValidationError(
                f"NumeroLinea 1-10000: {prefix}",
                [{"path": f"{prefix}.NumeroLinea", "code": "ecf.rango_numero_linea"}],
            )
        if not it.nombre_item or not it.nombre_item.strip():
            raise EcfValidationError(f"NombreItem requerido: {prefix}", [{"path": f"{prefix}.NombreItem", "code": "ecf.requerido"}])
        if it.nombre_item and len(it.nombre_item.strip()) > 80:
            raise EcfValidationError(f"NombreItem máximo 80: {prefix}", [{"path": f"{prefix}.NombreItem", "code": "ecf.largo_maximo"}])
        if it.indicador_facturacion not in indicadores_fact:
            raise EcfValidationError(
                f"IndicadorFacturacion 0-4: {prefix}",
                [{"path": f"{prefix}.IndicadorFacturacion", "code": "ecf.valor_no_permitido"}],
            )
        if tipo_ecf in (43, 44, 47) and it.indicador_facturacion != 4:
            raise EcfValidationError(
                f"Tipo {tipo_ecf} requiere IndicadorFacturacion 4: {prefix}",
                [{"path": f"{prefix}.IndicadorFacturacion", "code": "ecf.indicador_facturacion_tipo_ecf"}],
            )
        if tipo_ecf == 46 and it.indicador_facturacion != 3:
            raise EcfValidationError(
                f"Tipo 46 requiere IndicadorFacturacion 3: {prefix}",
                [{"path": f"{prefix}.IndicadorFacturacion", "code": "ecf.indicador_facturacion_tipo_ecf"}],
            )
        if it.indicador_bien_servicio not in indicadores_bien:
            raise EcfValidationError(
                f"IndicadorBienoServicio 1 o 2: {prefix}",
                [{"path": f"{prefix}.IndicadorBienoServicio", "code": "ecf.valor_no_permitido"}],
            )
        if tipo_ecf == 47 and it.indicador_bien_servicio != 2:
            raise EcfValidationError(
                f"Tipo 47 requiere IndicadorBienoServicio 2: {prefix}",
                [{"path": f"{prefix}.IndicadorBienoServicio", "code": "ecf.indicador_bien_servicio_tipo_47"}],
            )
        if it.cantidad <= 0:
            raise EcfValidationError(f"CantidadItem > 0: {prefix}", [{"path": f"{prefix}.CantidadItem", "code": "ecf.mayor_cero"}])
        if it.precio_unitario < 0 or it.monto_item < 0:
            raise EcfValidationError(
                f"PrecioUnitario/MontoItem no negativos: {prefix}",
                [{"path": f"{prefix}.MontoItem", "code": "ecf.valor_no_negativo"}],
            )


def validar_informacion_referencia(info: "InformacionReferencia") -> None:
    from .datos import InformacionReferencia

    if not info.ncf_modificado or not info.ncf_modificado.strip():
        raise EcfValidationError(
            "NCFModificado requerido para tipo 33/34",
            [{"path": "InformacionReferencia.NCFModificado", "code": "ecf.requerido"}],
        )
    validar_fecha("InformacionReferencia.FechaNCFModificado", info.fecha_ncf_modificado)
    if not info.codigo_modificacion or info.codigo_modificacion.strip() not in ("1", "2", "3", "4"):
        raise EcfValidationError(
            "CodigoModificacion debe ser 1, 2, 3 o 4",
            [{"path": "InformacionReferencia.CodigoModificacion", "code": "ecf.valor_no_permitido"}],
        )
